### PR TITLE
Added support for sorting folders above files.

### PR DESCRIPTION
**Pull request by cbrown1**
The only issue I am aware of is when building a full path to test whether a child is a folder. I don't really use lua much, and didn't find a function analogous to Python's os.path.join. So as you can see, I hardcoded something but I don't know how portable it is (I am on linux).

I tried to respect your effort to minimize if's inside loops, and the added functionality adds a layer of complexity when reading the code as a result. See what you think.

### CODE REVIEW

**Comment by sum01** [comment]
I don't actually see where this makes it show the full path on files. Am I missing something?

---

Anyways, you shouldn't need to make changes to the `get_scanlist()` function, as it already saves absolute paths into the `tree_view[]`.

`refresh_view()` would be a more applicable place to change the disaplayed content, as it uses `get_basename()` (which uses Golang's basname) on each filename when building the displayed file tree.

I suggest skipping the `get_basename()` call [here](https://github.com/NicolaiSoeborg/filemanager-plugin/blob/57df3ff47841916587986beaecace2295970a0f0/filemanager.lua#L308-L312) if you want to show the full paths.

I also personally think the `filemanager-foldersfirst` option should be `false` by default, as it's not the current default behaviour.

---

And for future knowledge, to join paths in a  portable way, you need to use `JoinPaths(dir, filename)`, although it's not needed for your proposed changes.

`JoinPaths()` is provided by the Micro plugin API, which internally calls Golang's join function.

**Comment by sum01** [comment]
I just now realized you meant this to show directires above files. My bad, I need some coffee...

**Comment by sum01** [comment]
What is the "problem opening files" you mentioned in the last commit?

**Comment by cbrown1** [comment]
I get the following error when I try to open a file (tab key):

``filemanager:563: attempt to call a non-function object
stack traceback:
	filemanager:563: in function 'try_open_at_y'
	filemanager:1157: in main chunk
	[G]: ?

Press enter to continue
``
Any ideas?

File: filemanager.lua [hunk 1]
```diff
@@ -180,39 +187,36 @@ local function get_scanlist(dir, ownership, indent_n)
...
 		end
-	else
-		-- Show dotfiles and ignored (aka everything)
-		for i = 1, #dir_scan do
-			results[i] = get_results_object(dir_scan[i]:Name())
+	end
+	if #files > 0 then
+	    -- Append any files to results, now that all folders have been added
+	    -- files will be > 0 only if folders_first and there are files
+		for i = 0, #files do
```
Comment (sum01) [inline filemanager.lua]: Index starts at 1 in Lua (unless you specifically create a zero index). This is (probably) what's causing the issues with `try_open_at_y`, as when it goes to access it, it gets a `nil` value.

There might even be a built-in "append table" type function in the Lua library, although I haven't looked at the reference manual in a while.

**Comment by sum01** [comment]
I would test this myself, but for some reason the nightly bin won't launch for me. If someone else is willing (@tommyshem @NicolaiSoeborg ?), pull his remote, fix the attempt to access a nil value, and see if this PR works :+1: 

Otherwise you'll be waiting on me to get Micro working.

**Comment by tommyshem** [comment]
Tested on Mac as is and it works. Not sure what the fix access nil value is?
Can you give me more info on how to cause the nil fault?
Does it need testing on Linux ?

To get the latest micro I just pulled it from GitHub and built it with `make build-all` and run from the go build directory.
My path is `~/go/src/github.com/zyedidia/micro`
run `./micro`


I don't use micro editor much now as using vscode with plugins but willing to help test with more info.

**Comment by sum01** [comment]
@tommyshem This https://github.com/NicolaiSoeborg/filemanager-plugin/pull/33#pullrequestreview-146636301

The fix is just changing the 0 to a 1.

And yeah, I haven't personally been using Micro for a while now either.

**Comment by tommyshem** [comment]
Tested on Mac 
micro --version
Version: 1.4.1
Commit hash: 1856891
Compiled on August 10, 2018

and nightly version 

./micro --version
Version: 1.4.2-dev.11
Commit hash: 6140dab
Compiled on August 21, 2018

I have not changed the 0 to 1 (original pull request) and I can not get it to crash with any errors it opens the right files. 

Tried in different directories and different files and can not get it to crash on Mac.

Is it related to #28 fault with micro 1.4.0

**Comment by sum01** [comment]
So I figured out how to get Micro worked, and debugged a little bit.

It turns out Lua is preventing an error here, which makes me curious why @cbrown1 even got errors.

---

This code does assign a `nil` value, but Lua "throws it away" since he used the length operator `#`, which stops on `nil` values. Thus the same index is assigned twice, since the first time it's `nil`, and the second time it will assign an actual "useful" filename/object.

**Pull request merged**